src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -208,6 +217,19 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> appl
                     // Call UpdateCompletedStreams() prior to frame processing in order to remove any streams that have exceeded their drain timeouts.
                     UpdateCompletedStreams();
 
+                    if (_keepAlive != null)
+                    {
+                        var state = _keepAlive.ProcessKeepAlive(!result.IsCanceled);
```

Comment:
Is this the right place to check for incoming data?
Is checking that IsCanceled is full on the ReadResult correct? My guess is no 😄 
Something we didn't consider is what should happen when there is flow control? That could impact receiving bytes and sending a ping, and also receiving the ACK to avoid timeout.

Reply:
I tested the different scenarios I could think of and this seems to work.

Reply:
> Is this the right place to check for incoming data?
> Is checking that IsCanceled is full on the ReadResult correct? My guess is no 😄 

This is more-or-less the right place, but instead of checking IsCanceled (which can be true even when there's also data leading to false negatives), I would change `_bytesReceivedCurrentTick` to `_frameReceivedCurrentTick` and only set it when TryReadFrame returns true.

> Something we didn't consider is what happens when there is flow control? That could impact receiving bytes and sending a ping, and also receiving the ACK to avoid timeout.

Fortunately, non-data frames are unaffected by flow control and a client cannot not even start attempting to send a data frame until there's a large enough connection and stream window for the entire frame.

Reply:
> I would change _bytesReceivedCurrentTick to _frameReceivedCurrentTick and only set it when TryReadFrame returns true.

Why do you want activity to only register when a complete frame is available? I didn't add a test for it, but I did think about the scenario of a large data frame (16KB) slowly arriving over a bad connection. It could take longer to receive that frame than the timeout.

Reply:
If it takes longer than 20s to receive 16kb then you're likely to start hitting your request body data rate limits anyways.

Reply:
It could be set to 1 second.

Reply:
If someone sets the timeout to 1 second, I think they'd be more likely to see aborted connections due to random latency spikes rather than low bandwidth. Though mybe that wouldn't be the case if that someone also configured Kestrel's max frame size to 16 MB instead of 16 KB.

We could look at the ReadResult's buffer size to see if it includes anything more than the unexamined data from the last read. Getting a RoS length shouldn't be too expensive.